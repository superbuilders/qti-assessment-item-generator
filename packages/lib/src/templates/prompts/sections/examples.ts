import { existsSync, readFileSync } from "node:fs"
import * as path from "node:path"
import { resolveLibPath } from "@/internal/paths"

const POSITIVE_EXAMPLE_CACHE = new Map<string, string>()
const NEGATIVE_EXAMPLE_CACHE = new Map<string, string>()

const POSITIVE_ROOT = resolveLibPath("templates/prompts/examples/positive")
const POSITIVE_TEMPLATES_DIR = path.join(POSITIVE_ROOT, "templates")
const POSITIVE_NOTES_DIR = path.join(POSITIVE_ROOT, "notes")

const NEGATIVE_ROOT = resolveLibPath("templates/prompts/examples/negative")
const NEGATIVE_TEMPLATES_DIR = path.join(NEGATIVE_ROOT, "templates")
const NEGATIVE_NOTES_DIR = path.join(NEGATIVE_ROOT, "notes")

export function createPositiveExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadPositiveExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### POSITIVE_EXAMPLES
<examples type="positive">
${rendered}
</examples>`
}

export function createNegativeExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadNegativeExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### NEGATIVE_EXAMPLES
<examples type="negative">
${rendered}
</examples>`
}

function loadPositiveExample(name: string): string {
	const cached = POSITIVE_EXAMPLE_CACHE.get(name)
	if (cached) return cached

	const codePath = path.join(POSITIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}

	const notesPath = path.join(POSITIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""

	const rendered = `<example kind="positive" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	POSITIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}

function loadNegativeExample(name: string): string {
	const existing = NEGATIVE_EXAMPLE_CACHE.get(name)
	if (existing) return existing

	const codePath = path.join(NEGATIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}
	const notesPath = path.join(NEGATIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""
	const rendered = `<example kind="negative" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	NEGATIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}
