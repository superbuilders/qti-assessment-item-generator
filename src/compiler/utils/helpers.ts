import { typedSchemas } from "@/widgets/registry"

/**
 * Calculates the greatest common divisor (GCD) of two numbers using the Euclidean algorithm.
 * @param a The first number.
 * @param b The second number.
 * @returns The GCD of a and b.
 */
export function gcd(a: number, b: number): number {
	return b === 0 ? a : gcd(b, a % b)
}

/**
 * Determines if a fraction results in a terminating decimal.
 * A fraction terminates if and only if the prime factors of its simplified denominator are only 2s and 5s.
 * @param numerator The numerator of the fraction.
 * @param denominator The denominator of the fraction.
 * @returns True if the fraction terminates, false otherwise.
 */
export function isTerminatingFraction(
	numerator: number,
	denominator: number
): boolean {
	if (denominator === 0) {
		return false
	}
	// Simplify the fraction by dividing by the GCD.
	const commonDivisor = gcd(Math.abs(numerator), Math.abs(denominator))
	let simplifiedDen = Math.abs(denominator) / commonDivisor

	// The fraction terminates if the simplified denominator's prime factors are only 2 and 5.
	while (simplifiedDen % 2 === 0) {
		simplifiedDen /= 2
	}
	while (simplifiedDen % 5 === 0) {
		simplifiedDen /= 5
	}

	return simplifiedDen === 1
}

export function encodeDataUri(content: string): string {
	const base64 = btoa(unescape(encodeURIComponent(content)))
	const isSvg = content.trim().startsWith("<svg")
	return `${isSvg ? "data:image/svg+xml" : "data:text/html"};base64,${base64}`
}

// Helper function to check if a string is a valid widget type
export function isValidWidgetType(
	type: string
): type is keyof typeof typedSchemas {
	return type in typedSchemas
}

/**
 * Attempts to approximate simple multiplicative/divisive expressions that include π or pi
 * by replacing pi with the provided approximation (default 3.14) and evaluating the expression.
 *
 * Supported forms (case-insensitive for "pi"):
 * - "144*pi", "pi*144", "144pi" (implicit multiplication), "pi144" (rare, but treated as pi*144)
 * - Chains with '*' and '/' only, e.g. "3/2*pi", "pi/4", "6*pi/5", "(144) * (pi)"
 * Parentheses are allowed but only for grouping; no addition/subtraction is supported.
 *
 * Returns a normalized decimal string if evaluation succeeds, otherwise null.
 */
export function tryApproximatePiProduct(
	rawExpr: string,
	piApprox = 3.14
): string | null {
	if (typeof rawExpr !== "string" || rawExpr.trim() === "") return null

	// Quick check: must contain a pi token of some sort
	const containsPi = /\bpi\b/i.test(rawExpr) || rawExpr.includes("π")
	if (!containsPi) return null

	// Normalize: insert explicit * for implicit numeric-pi adjacency in common cases
	const expr = rawExpr
		// insert * between number and pi (e.g., 144pi -> 144*pi)
		.replace(/(\d)\s*(π|\bpi\b)/gi, "$1*$2")
		// and between pi and number (e.g., pi144 -> pi*144)
		.replace(/(π|\bpi\b)\s*(\d)/gi, "$1*$2")
		// remove whitespace
		.replace(/\s+/g, "")
		// drop parentheses (multiplicative-only evaluation does not require them)
		.replace(/[()]/g, "")
		// normalize 'pi' variants to numeric approximation
		.replace(/π/gi, String(piApprox))
		.replace(/\bpi\b/gi, String(piApprox))

	// After substitution, we only allow digits, '.', '*', '/' characters
	if (/[^0-9.*/]/.test(expr)) return null
	// Disallow addition/subtraction in this lightweight evaluator
	if (/[+-]/.test(expr)) return null
	// Must still contain the approximation value to ensure we actually used pi
	if (!expr.includes(String(piApprox))) return null

	// Split into tokens by * and / and evaluate left-to-right (* and / are same precedence)
	const parts: string[] = expr.split(/([*/])/)
	if (parts.length === 0) return null

	// Validate tokens
	for (let i = 0; i < parts.length; i++) {
		const t = parts[i]
		if (i % 2 === 0) {
			// operand
			if (t === "" || Number.isNaN(Number(t))) return null
		} else {
			// operator must be * or /
			if (t !== "*" && t !== "/") return null
		}
	}

	let acc = Number(parts[0])
	for (let i = 1; i < parts.length; i += 2) {
		const op = parts[i]
		const rhs = Number(parts[i + 1])
		if (op === "*") acc = acc * rhs
		else {
			if (rhs === 0) return null
			acc = acc / rhs
		}
	}

	// Normalize number to a compact string while preserving significant digits
	// Avoid scientific notation for typical classroom values
	const normalizeNumberString = () => {
		const maybeFixed = acc.toFixed(10) // limit noise
		// trim trailing zeros
		const trimmed = maybeFixed.replace(/\.0+$/, "").replace(/(\.\d*?)0+$/, "$1")
		return trimmed
	}
	const asString = normalizeNumberString()

	// Add a guard: must be a finite number
	if (!Number.isFinite(Number(asString))) return null
	return asString
}

/**
 * Normalizes a string part for use in a derived QTI identifier.
 * Converts to upper snake case and removes invalid characters.
 * @param part The string part to normalize.
 * @returns The normalized string.
 */
export function normalizeIdPart(part: string): string {
	return part.toUpperCase().replace(/[^A-Z0-9_]/g, "_")
}

/**
 * Derives a deterministic, QTI-safe combo identifier from a path of response parts.
 * @param pathParts An array of strings, where each string is a response-value pair (e.g., "RESPONSE_1_A").
 * @returns The fully-formed combo identifier (e.g., "FB__RESPONSE_1_A__RESPONSE_2_CORRECT").
 */
export function deriveComboIdentifier(pathParts: string[]): string {
	return `FB__${pathParts.join("__")}`
}
