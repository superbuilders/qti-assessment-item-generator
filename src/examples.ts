import type { AssessmentItemInput } from "@/core/item"
import type { allWidgetsCollection } from "@/widgets/collections/all"
import type { WidgetTypeTupleFrom } from "@/widgets/collections/types"

export const allExamples: AssessmentItemInput<WidgetTypeTupleFrom<typeof allWidgetsCollection>>[] = []
