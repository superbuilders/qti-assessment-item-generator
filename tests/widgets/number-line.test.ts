import { describe, expect, test } from "bun:test"
import * as errors from "@superbuilders/errors"
import * as logger from "@superbuilders/slog"
import { numberLineExamples } from "@/examples/number-line"
import { generateWidgetForTest } from "@/testing/helpers/generateWidgetForTest"
import type { WidgetInput } from "@/widgets/registry"

describe("Widget: number-line", () => {
	const examples: WidgetInput[] = numberLineExamples

	examples.forEach((props, index) => {
		test(`should produce consistent output for example #${index + 1}`, async () => {
			const result = await errors.try(generateWidgetForTest(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error, index })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})
	})
})
