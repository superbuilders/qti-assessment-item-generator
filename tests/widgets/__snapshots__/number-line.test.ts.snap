// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Widget: number-line should produce consistent output for example #1 1`] = `"<svg width="540" height="160" viewBox="-20 -20 540 160" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><marker id="action-arrow" viewBox="0 0 10 10" refX="8" refY="5" markerWidth="6" markerHeight="6" orient="auto"><path d="M 0 0 L 10 5 L 0 10 z" fill="#333333"/></marker></defs><line x1="20" y1="60" x2="480" y2="60" stroke="#333333" stroke-width="2"/><line x1="20" y1="55" x2="20" y2="65" stroke="#333333" stroke-width="1.5"/><text x="20" y="80" font-size="11" text-anchor="middle" fill="#333333">0</text><line x1="66" y1="55" x2="66" y2="65" stroke="#333333" stroke-width="1.5"/><text x="66" y="80" font-size="11" text-anchor="middle" fill="#333333">1</text><line x1="112" y1="55" x2="112" y2="65" stroke="#333333" stroke-width="1.5"/><text x="112" y="80" font-size="11" text-anchor="middle" fill="#333333">2</text><line x1="158" y1="55" x2="158" y2="65" stroke="#333333" stroke-width="1.5"/><text x="158" y="80" font-size="11" text-anchor="middle" fill="#333333">3</text><line x1="204" y1="55" x2="204" y2="65" stroke="#333333" stroke-width="1.5"/><text x="204" y="80" font-size="11" text-anchor="middle" fill="#333333">4</text><line x1="250" y1="55" x2="250" y2="65" stroke="#333333" stroke-width="1.5"/><text x="250" y="80" font-size="11" text-anchor="middle" fill="#333333">5</text><line x1="296" y1="55" x2="296" y2="65" stroke="#333333" stroke-width="1.5"/><text x="296" y="80" font-size="11" text-anchor="middle" fill="#333333">6</text><line x1="342" y1="55" x2="342" y2="65" stroke="#333333" stroke-width="1.5"/><text x="342" y="80" font-size="11" text-anchor="middle" fill="#333333">7</text><line x1="388" y1="55" x2="388" y2="65" stroke="#333333" stroke-width="1.5"/><text x="388" y="80" font-size="11" text-anchor="middle" fill="#333333">8</text><line x1="434" y1="55" x2="434" y2="65" stroke="#333333" stroke-width="1.5"/><text x="434" y="80" font-size="11" text-anchor="middle" fill="#333333">9</text><line x1="480" y1="55" x2="480" y2="65" stroke="#333333" stroke-width="1.5"/><text x="480" y="80" font-size="11" text-anchor="middle" fill="#333333">10</text><circle cx="158" cy="60" r="5" fill="#4285F4" stroke="#4285F4" stroke-width="1"/><line x1="158" y1="35" x2="158" y2="52" stroke="#333333" stroke-width="2" marker-end="url(#action-arrow)"/><text x="158" y="30" font-size="12" text-anchor="middle" font-weight="700" fill="#4285F4">3</text><circle cx="365" cy="60" r="5" fill="#FF6B6B" stroke="#FF6B6B" stroke-width="1"/><text x="365" y="45" font-size="12" text-anchor="middle" font-weight="700" fill="#FF6B6B">7.5</text></svg>"`;

exports[`Widget: number-line should produce consistent output for example #2 1`] = `"<svg width="440" height="140" viewBox="-20 -20 440 140" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><marker id="action-arrow" viewBox="0 0 10 10" refX="8" refY="5" markerWidth="6" markerHeight="6" orient="auto"><path d="M 0 0 L 10 5 L 0 10 z" fill="#333333"/></marker></defs><line x1="20" y1="50" x2="380" y2="50" stroke="#333333" stroke-width="2"/><line x1="20" y1="45" x2="20" y2="55" stroke="#333333" stroke-width="1.5"/><text x="20" y="70" font-size="11" text-anchor="middle" fill="#333333">-5</text><line x1="56" y1="45" x2="56" y2="55" stroke="#333333" stroke-width="1.5"/><text x="56" y="70" font-size="11" text-anchor="middle" fill="#333333">-4</text><line x1="92" y1="45" x2="92" y2="55" stroke="#333333" stroke-width="1.5"/><text x="92" y="70" font-size="11" text-anchor="middle" fill="#333333">-3</text><line x1="128" y1="45" x2="128" y2="55" stroke="#333333" stroke-width="1.5"/><text x="128" y="70" font-size="11" text-anchor="middle" fill="#333333">-2</text><line x1="164" y1="45" x2="164" y2="55" stroke="#333333" stroke-width="1.5"/><text x="164" y="70" font-size="11" text-anchor="middle" fill="#333333">-1</text><line x1="200" y1="45" x2="200" y2="55" stroke="#333333" stroke-width="1.5"/><text x="200" y="70" font-size="11" text-anchor="middle" fill="#333333">0</text><line x1="236" y1="45" x2="236" y2="55" stroke="#333333" stroke-width="1.5"/><text x="236" y="70" font-size="11" text-anchor="middle" fill="#333333">1</text><line x1="272" y1="45" x2="272" y2="55" stroke="#333333" stroke-width="1.5"/><text x="272" y="70" font-size="11" text-anchor="middle" fill="#333333">2</text><line x1="308" y1="45" x2="308" y2="55" stroke="#333333" stroke-width="1.5"/><text x="308" y="70" font-size="11" text-anchor="middle" fill="#333333">3</text><line x1="344" y1="45" x2="344" y2="55" stroke="#333333" stroke-width="1.5"/><text x="344" y="70" font-size="11" text-anchor="middle" fill="#333333">4</text><line x1="380" y1="45" x2="380" y2="55" stroke="#333333" stroke-width="1.5"/><text x="380" y="70" font-size="11" text-anchor="middle" fill="#333333">5</text><circle cx="128" cy="50" r="5" fill="#34A853" stroke="#34A853" stroke-width="1"/><line x1="128" y1="25" x2="128" y2="42" stroke="#333333" stroke-width="2" marker-end="url(#action-arrow)"/><text x="128" y="20" font-size="12" text-anchor="middle" font-weight="700" fill="#34A853">-2</text><circle cx="344" cy="50" r="5" fill="#FF6B6B" stroke="#FF6B6B" stroke-width="1"/><line x1="344" y1="25" x2="344" y2="42" stroke="#333333" stroke-width="2" marker-end="url(#action-arrow)"/><text x="344" y="20" font-size="12" text-anchor="middle" font-weight="700" fill="#FF6B6B">4</text></svg>"`;

exports[`Widget: number-line should produce consistent output for example #3 1`] = `"<svg width="560" height="160" viewBox="-20 -20 560 160" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><marker id="action-arrow" viewBox="0 0 10 10" refX="8" refY="5" markerWidth="6" markerHeight="6" orient="auto"><path d="M 0 0 L 10 5 L 0 10 z" fill="#333333"/></marker></defs><line x1="20" y1="60" x2="500" y2="60" stroke="#333333" stroke-width="2"/><line x1="32" y1="57" x2="32" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="44" y1="57" x2="44" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="56" y1="57" x2="56" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="68" y1="57" x2="68" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="92" y1="57" x2="92" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="104" y1="57" x2="104" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="116" y1="57" x2="116" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="128" y1="57" x2="128" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="152" y1="57" x2="152" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="164" y1="57" x2="164" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="176" y1="57" x2="176" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="188" y1="57" x2="188" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="212" y1="57" x2="212" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="224" y1="57" x2="224" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="236" y1="57" x2="236" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="248" y1="57" x2="248" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="272" y1="57" x2="272" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="284" y1="57" x2="284" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="296" y1="57" x2="296" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="308" y1="57" x2="308" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="332" y1="57" x2="332" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="344" y1="57" x2="344" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="356" y1="57" x2="356" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="368" y1="57" x2="368" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="392" y1="57" x2="392" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="404" y1="57" x2="404" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="416" y1="57" x2="416" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="428" y1="57" x2="428" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="452" y1="57" x2="452" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="464" y1="57" x2="464" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="476" y1="57" x2="476" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="488" y1="57" x2="488" y2="63" stroke="#333333" stroke-width="1.5"/><line x1="20" y1="55" x2="20" y2="65" stroke="#333333" stroke-width="1.5"/><line x1="80" y1="55" x2="80" y2="65" stroke="#333333" stroke-width="1.5"/><line x1="140" y1="55" x2="140" y2="65" stroke="#333333" stroke-width="1.5"/><line x1="200" y1="55" x2="200" y2="65" stroke="#333333" stroke-width="1.5"/><line x1="260" y1="55" x2="260" y2="65" stroke="#333333" stroke-width="1.5"/><line x1="320" y1="55" x2="320" y2="65" stroke="#333333" stroke-width="1.5"/><line x1="380" y1="55" x2="380" y2="65" stroke="#333333" stroke-width="1.5"/><line x1="440" y1="55" x2="440" y2="65" stroke="#333333" stroke-width="1.5"/><line x1="500" y1="55" x2="500" y2="65" stroke="#333333" stroke-width="1.5"/><circle cx="212" cy="60" r="5" fill="#1E40AF" stroke="#1E40AF" stroke-width="1"/><circle cx="320" cy="60" r="5" fill="#1E40AF" stroke="#1E40AF" stroke-width="1"/><text x="320" y="45" font-size="12" text-anchor="middle" font-weight="700" fill="#1E40AF">125</text><circle cx="164" cy="60" r="5" fill="#111111" stroke="#111111" stroke-width="1"/><rect x="158" y="39" width="12" height="12" fill="#ffffff" stroke="#111111" stroke-width="1.5"/></svg>"`;
